Multiwallet Feature:  Generate Seed from Dice Rolls

tdb3 opened the pull request:
Added a convenience feature to multiwallet, guiding a user through seed creation from dice rolls.
The feature can be accessed through the menu option ```generate_seed_from_dice```.
The test cases are based on Coinkite's examples (https://coldcard.com/docs/verifying-dice-roll-math).

Happy to tweak as needed.

mflaxman commented:
Very cool, thank you for the PR @tdb3! I'm partial to pulling words out of a hat (vs rolling dice) as it's one less step to verify (dice->words->xpub vs just words->xpub), but I like having support for both as long as it's done in a way that matches other implementations.

I think this should be part of the `HDPrivateKey` class, perhaps a new `HDPrivateKey.from_dice()` method? That way it's accessible to developers using buidl (not just multiwallet users) and also so it can have bulletproof unit testing (the testing harness for the CLI apps is admittedly a bit janky). What do you think of moving it there?

tdb3 commented:
Sounds good.  How about something like the following?

Since ```HDPrivateKey.from_dice()``` is essentially a specific example of creating a seed from bytes, a more general ```HDPrivateKey.from_bytes()``` is created, which ```from_dice()``` calls.  ```HDPrivateKey.from_bytes()``` incorporates similar arguments (e.g., optional password, etc.) as ```HDPrivateKey.from_mnemonic()```, keeps the HMAC SHA512 approach (for integration of the password as salt, and stretching), and returns the HDPrivateKey as well as the associated mnemonic words generated from the bytes provided (for later re-use).

mflaxman commented:
I'm not 100% sure I'm following what you're saying, maybe it would be easier if you wrote it as a PR and then it will be easy to comment on the code?

My concern with a `HDPrivateKey.from_bytes()` is developer ergonomics. `buidl` is very cautious about making it hard for regular uesrs to footgun, and only experts should go outside commonly accepted standards (i.e. bip39 mnemonics). If there's a need for a `from_bytes()` method, it should at least be a private method `_from_bytes()`, although python doesn't enforce that it's at least provides some signaling.

tdb3 commented:
Rebased, modified approach to address comments from @mflaxman, force pushed.

mflaxman commented on buidl/test/test_hd.py:
```diff
@@ -408,6 +413,30 @@ def test_from_mnemonic(self):
+            [
+                "123456123456123456123456123456123456123456123456123456123456123456123456123456123456123456123456123456",
+                "more matter caught bind tip twin indicate visa rifle angle defense lizard stock cave cradle injury always mule photo horse range opinion affair garlic",
+                "xprv9s21ZrQH143K4J7isGyg4whNPAMGdfNntamxzPk2qFDS4nNJtNwrZ6vaiv6Jk5JsNyXKfZT5X3QnEMaTg3uvcKszhdgbLxgMh6ETE3nrzPn",
```
Is this what coldcard outputs?

mflaxman replied:
Can you add [this test vector](https://www.keepitsimplebitcoin.com/coldcards-dice-roll-math/) for `523365252662366`? It was the only other one I could find online.

tdb3 replied:
Sure thing.  Added in 6de06301bbfc795e8474bca57d1d3b019c5c8da0.

tdb3 commented:
Force pushed to address linter error.

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -98,6 +98,12 @@ def bytes_to_mnemonic(b, num_bits):
 
+def dice_rolls_to_mnemonic(dice_rolls):
+    """returns a mnemonic from a string of dice rolls (>=99 rolls recommended)"""
+    rolls_hash = sha256(dice_rolls.encode())
```
Let's add a sanity check that `dice_rolls` is a string containing only the numbers 1-6. Anything else and it should throw an error.

tdb3 replied:
Agreed.  Added in 6de06301bbfc795e8474bca57d1d3b019c5c8da0.

mflaxman commented on buidl/hd.py:
```diff
@@ -480,6 +481,22 @@ def from_mnemonic(
+        priv_version=None,
+        pub_version=None,
+    ):
+        """Returns an HDPrivateKey object from a string of dice rolls (>=99 rolls recommended)"""
```
We need some protection for non-cryptographically inclined developers. [From coldcard](https://coldcard.com/docs/verifying-dice-roll-math/"), "for 128-bit security, which we consider the absolute minimum, you need 50 rolls, and for 256-bits of security, 99 rolls"

I think the best move here is to add a keyword argument `allow_low_entropy=False` to this method signature. If `True`, then you can have entropy below the desired limit. If `False`, then you must have 50-99 rolls (depending on mnemonic length) or it will throw an error.

mflaxman commented:
Thanks for the PR @tdb3!

Why support only 24 words? It's the default/best choice, but I know lots of people with 12 word seed phrases that use passphrases for extra entropy. I don't think this is a requirement for merging, I'd just think you'd want to support it since you already did most of the work.

I added two small comments, overall nicely done!

tdb3 commented:
Great comments.
Comments addressed inline.

Updated to support 12 through 24 words.

Overall, decided to simplify things.  `HDPrivateKey.from_dice()` isn't an absolute necessity when `dice_rolls_to_mnemonic()` exists.  `dice_rolls_to_mnemonic()` is sufficient to provide the mnemonic words that can then be used with HDPrivateKey.from_mnemonic().  This results in a smaller code footprint.

Added additional test cases for 12-word (and other) entries, and edge cases (such as only using values 1-6, string provided, minimum entropy checking).  These were double-checked against https://iancoleman.io/bip39/ and https://jlopp.github.io/xpub-converter/

Multiwallet receive raised exceptions from `dice_rolls_to_mnemonic()` when improper data (e.g. wrong number of words, insufficient entropy) is input.  This way applications other than Multiwallet receive similar error handling.

Below is the output from Coinkite's rolls12.py script for comparison (these are baked into the unit tests):
```
$ python3 rolls12.py 

e3b0c44298fc1c149afbf4c8996fb924

WARNING: Input is empty. This is a known wallet

WARNING: Input is only 0 bits of entropy

   1: together
   2: mail
   3: awful
   4: cradle
   5: scrub
   6: apart
   7: hip
   8: leader
   9: silk
  10: slice
  11: unusual
  12: embark


$ python3 rolls12.py 
123456
8d969eef6ecad3c29a3a629280e686cf

WARNING: Input is only 15 bits of entropy

   1: mirror
   2: reject
   3: rookie
   4: talk
   5: pudding
   6: throw
   7: happy
   8: era
   9: myth
  10: already
  11: payment
  12: owner


$ python3 rolls12.py 
12345612345612345612345612345612345612345612345612
ee72ae915a4e6ea7ccbeb8e5e5eecef2

   1: unveil
   2: nice
   3: picture
   4: region
   5: tragic
   6: fault
   7: cream
   8: strike
   9: tourist
  10: control
  11: recipe
  12: tourist
```

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -98,6 +99,40 @@ def bytes_to_mnemonic(b, num_bits):
+    if not isinstance(dice_rolls, str):
+        raise ValueError("Dice rolls must be provided as a string")
+    if (
+        len([roll for roll in dice_rolls if roll not in ("1", "2", "3", "4", "5", "6")])
```
Nit: this is fine, but can be made simpler (and with a clearer error message) without the list comprehension.

```
valid_dice_rolls = set(["1", "2", "3", "4", "5", "6"])
for roll in dice_rolls:
    if roll not in valid_dice_rolls:
        raise ValueError(f"Invalid dice roll: {roll}")
```

mflaxman commented on multiwallet.py:
```diff
@@ -246,6 +246,35 @@ def _get_bip39_firstwords():
+        maximum=24,
+    )
+    if num_words not in (12, 15, 18, 21, 24):
+        raise ValueError(
```
This entire method needs a `while True` block, otherwise providing an invalid entry will break (good), but the UI of that for the end-user starts them over (bad).

mflaxman commented on multiwallet.py:
```diff
@@ -246,6 +246,35 @@ def _get_bip39_firstwords():
+            '\n(e.g., "14625363...", >= 100 values recommended): '
+        )
+    ).strip()
+    return dice_vals
```
This needs to handle bad entries (like someone entering a letter, space, a number not 1-6, etc) gracefully

mflaxman commented on multiwallet.py:
```diff
@@ -487,6 +516,55 @@ def do_generate_seed(self, arg):
+        print_yellow(f"\nPUBLIC KEY INFO ({network})")
+        print_yellow("Copy-paste this into Specter-Desktop:")
+        print_green(key_record)
+
```
This has a lot of overlap with `do_generate_seed()`, what do you think about refactoring the code so that the shared code only lives in one place? For example, I noticed that yours doesn't include the footgun warning that is in `do_generate_seed`. DRYing it out would seem to make more sense.

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -98,6 +99,40 @@ def bytes_to_mnemonic(b, num_bits):
+    entropy_per_roll = 2.585
+    mnemonic_entropy = {12: 128, 15: 160, 18: 192, 21: 224, 24: 256}
+    # check that the appropriate amount of entropy had been provided
+    min_rolls_needed = ceil(mnemonic_entropy[num_words] / entropy_per_roll)
```
While technically correct (and great to see the code on how it's done), this would be easier to read for developers who are not cryptographically inclined if you included the calculation results in the `mnemonic_entropy` dict above. Otherwise, if you get an error for an where `num_words` is `12` you don't know how many rolls are needed for another entry where `num_words` is 24.

Something like
```
entropy_calcs_dict = {
    # num_words: [bits_of_entropy, min_rolls_needed],
    12: [128, 50],
    15: [160, 62],
    18: [192, 75],
    21: [224, 87],
    24: [256, 100],
}
# min_rolls_needed is calculated as ceil(num_words / entropy_per_roll), where entropy_per_roll is log2(6)=2.585
```

mflaxman reviewed (commented):
This is a great PR @tdb3. If I were still working, I'd be trying to hire you!

Added some minor things. I'm holding this PR to such a high standard because it touches private key generation and exposes that to less-technical users (multiwallet), so developer ergonomics and UX are paramount.

mflaxman commented:
Also, this is a substantial PR! Let's bump the version number in `setup.py`.
